Refresh the august access token when needed

bdraco opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


Currently august will stop working when the token
expires about every six month.

This fixes #23788

Also in review #31732 (please review this one first)
Next up:
* Some general refactoring and tests (since this module needs a bit of love in that area)
* Fixes for the doorbell code (I had to buy one before could start working through these issues)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #23788
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

balloob commented:
Nice work on all this. Can you please add yourself as a code owner in the manifest and run `python3 -m script.hassfest`.

balloob commented on homeassistant/components/august/__init__.py:
```diff
@@ -226,6 +232,29 @@ def locks(self):
+    def refresh_access_token_if_needed(self):
+        """Refresh the august access token if needed."""
+
+        if self._authenticator.should_refresh():
```
I don't see a lock here, which means that you can fetch multiple new tokens at the same time.

bdraco replied:
@balloob This probably needs a look.  Which locking mechanism is recommended?

balloob replied:
We only use async locks because normal locks can starve our thread pool. 

In async land we cannot do I/O as that halts the system. You can run functions that do I/O in async world with `await hass.async_add_executor_job(function, arg1, arg2)`. Functions `lock`, `unlock` and `update` need to be updated to be async `async def async_X`

bdraco replied:
Since this can come in from doorbell, camera, or lock it looks like I'd have to convert all of these to async.   I'm still learning the codebase so its a bit outside of my comfort zone.  Since we know we have to call the activity api, I'm going to move the token refresh there since it will already be blocked from happening at the same time by the Throttle

balloob commented on homeassistant/components/august/lock.py:
```diff
@@ -15,6 +15,8 @@
 
 SCAN_INTERVAL = timedelta(seconds=10)
 
+PARALLEL_UPDATES = 0
```
Setting parallel updates to 0 means that all requests for all entities are happening at the same time. State updates but also requests to lock/unlock locks. Can the API handle that ?

bdraco replied:
Just found _get_parallel_updates_semaphore.  I think it should actually be PARALLEL_UPDATES = 1 so only one entity can update at a time

balloob replied:
1 is the default.

bdraco commented:
> Nice work on all this. Can you please add yourself as a code owner in the manifest and run `python3 -m script.hassfest`.

Will do in the next resubmit

bdraco committed:
Refresh the august access token
when needed.

Currently august will stop working when the token
expires about every six month.

This resolves issue #23788

bdraco commented:
Apologizes for the extra commits.  I'm still getting used to the github workflow.

bdraco committed:
Make refresh_access_token_if_needed private since we do not want additional callers

bdraco committed:
Add init

balloob reviewed (approved):
🎉 nice!
